Deduplicate bibdata playbooks

acozine opened the pull request:
Consolidates the bibdata production and staging playbooks. This will need additional changes once we have QA environments. I'll rebase as necessary. 

Related to #2618 and #2616.

acozine commented:
Rebased and then updated to incorporate the QA environment as well as production and staging.

carolyncole commented on playbooks/bibdata.yml:
```diff
@@ -0,0 +1,26 @@
+    - role: roles/bibdata
+    - role: roles/bibdata_sqs_poller
+    - role: roles/sidekiq_worker
+      when: "'worker' in inventory_hostname"
```
Cool when!

acozine replied:
I don't know if production needs the `samba_shares_root` var - so far I've only moved what existed before. But it certainly looks odd to have that var in staging and QA and not in production.

acozine replied:
Ah, it does need it. I missed this; it was in the production playbook originally.

carolyncole reviewed (commented):
@acozine just one question.  Does production need the `samba_shares_root` variable defined too?

carolyncole reviewed (changes requested):
@acozine I get an error when I run the playbook
```
$ ansible-playbook playbooks/bibdata.yml -e runtime_env=qa
ERROR! We were unable to read either as JSON nor YAML, these are the errors we got from each:
JSON: Expecting value: line 1 column 1 (char 0)

Syntax Error while loading YAML.
  mapping values are not allowed in this context

The error appears to be in '/Users/colec/projects/princeton_ansible/playbooks/bibdata.yml': line 6, column 8, but may
be elsewhere in the file depending on the exact syntax problem.

The offending line appears to be:

- Build bibdata infrastructure
  hosts: bibdata_{{ runtime_env | default('staging') }}
       ^ here
We could be wrong, but this one looks like it might be an issue with
missing quotes. Always quote template expression brackets when they
start a value. For instance:

    with_items:
      - {{ foo }}

Should be written as:

    with_items:
      - "{{ foo }}"
```

carolyncole commented on playbooks/bibdata.yml:
```diff
@@ -0,0 +1,26 @@
+# by default this playbook runs in the staging environment
+# to run in qa, pass '-e runtime_env=qa'
+# to run in production, pass '-e runtime_env=production'
+- Build bibdata infrastructure
```
Needs `name:` to fix the syntax error issue I saw and commented on earlier

carolyncole replied:
Although once this change is in then the playbook fails on 
```
TASK [postgresql : PostgreSQL | ensure access to postgres server for client connections] *************************************************************
fatal: [bibdata-qa1.princeton.edu -> lib-postgres3.princeton.edu]: FAILED! => {"changed": false, "msg": "Destination /etc/postgresql/13/main/pg_hba.conf does not exist !", "rc": 257}
fatal: [bibdata-worker-qa1.princeton.edu -> lib-postgres3.princeton.edu]: FAILED! => {"changed": false, "msg": "Destination /etc/postgresql/13/main/pg_hba.conf does not exist !", "rc": 257}

PLAY RECAP *******************************************************************************************************************************************
bibdata-qa1.princeton.edu  : ok=49   changed=3    unreachable=0    failed=1    skipped=19   rescued=0    ignored=0   
bibdata-worker-qa1.princeton.edu : ok=49   changed=3    unreachable=0    failed=1    skipped=19   rescued=0    ignored=0   
```

acozine replied:
D'oh! I fixed the syntax thing by adding `name:`, thanks for catching that @carolyncole.

I'll do some digging on the postgresql issue.

acozine replied:
Okay, the postgresql failure is a legit error. On the `lib-postgres3` box the installed version of pgsql is 10:
```
postgres@lib-postgres3:/home/pulsys$ ls -la /etc/postgresql/
total 20
drwxr-xr-x   3 postgres postgres  4096 Mar 15  2018 .
drwxr-xr-x 116 root     root     12288 Mar  8 15:22 ..
drwxr-xr-x   3 postgres postgres  4096 Mar 15  2018 10
postgres@lib-postgres3:/home/pulsys$ ls -la /etc/postgresql/10/main/
total 68
drwxr-xr-x 3 postgres postgres  4096 Mar 24 11:16 .
drwxr-xr-x 3 postgres postgres  4096 Mar 15  2018 ..
drwxr-xr-x 2 postgres postgres  4096 Jul 15  2021 conf.d
-rw-r--r-- 1 postgres postgres   315 Mar 15  2018 environment
-rw-r--r-- 1 postgres postgres   143 Mar 15  2018 pg_ctl.conf
-rw-r----- 1 postgres postgres  9968 Mar 24 11:16 pg_hba.conf
-rw-r----- 1 postgres postgres  1636 Mar 15  2018 pg_ident.conf
-rw-r--r-- 1 postgres postgres 22538 Sep 28  2020 postgresql.conf
-rw-r--r-- 1 root     root       681 Mar 17  2018 repmgr.conf
-rw-r--r-- 1 postgres postgres   317 Mar 15  2018 start.conf
```
Should the bibdata boxes be looking at a different pgsql server?

acozine commented:
Ready for (re)-review @carolyncole. 

I fixed the syntax problem, updated the QA vars to look at the `lib-postgres-staging3` server instead of `lib-postgres3`, and looked into the `samba_shares_root` question. The staging and production bibdata servers both have the same symlinks in the `/data` directory, so I think we need the variable (and the same value) in all environments.

acozine commented:
The error on the `PostgreSQL | create postgresql database` task is happening because the posgresql role is not creating the correct `pg_hba.conf` entry for this database. The task (`PostgreSQL | ensure access to postgres server for client connections`) uses this formula:
`host  {{ pg_hba_postgresql_database }}      {{ pg_hba_postgresql_user }} {{ ansible_default_ipv4.address }}/32       md5`
and on the staging machine that is currently creating this line:
`host  dbname      pguser ##(ip redacted)##       md5`
which reflects the [default values](https://github.com/pulibrary/princeton_ansible/blob/383f5cb59deb5cb0dc7f899fdd87eba2fa0c6f07/roles/postgresql/defaults/main.yml#L75-L76) in the postgresql role.

acozine commented:
Fixing the error reported above required changes to the postgresql role. I've put those changes into a separate PR, so this PR is now blocked until we review and (I hope) merge #2902.

acozine commented:
Rebased with the changes from #2902. If the tests pass, this PR is ready for final review @carolyncole.

carolyncole commented on playbooks/bibdata.yml:
```diff
@@ -0,0 +1,26 @@
+      when: "'worker' in inventory_hostname"
+    - role: 'hr_share'
+    - role: roles/datadog
+      when: runtime_env == "production"
```
@acozine I really hate to delay this PR any longer, but I noticed yesterday that this line causes issue when runtime_env is not set in the approvals playbook.  I created a PR for approvals to fix it (#2910)  Would you prefer to merge this PR and ticket the fact that it will not run without the runtime_env being set, or did you want to update it prior to me merging?

carolyncole commented:
@acozine something is still not quite right in the postgres setup for this PR:  
```
TASK [postgresql : PostgreSQL | create postgresql database] *******************************************************************
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: connection to server at "lib-postgres-staging3.princeton.edu" (128.112.204.72), port 5432 failed: FATAL:  password authentication failed for user "postgresadmin"
fatal: [bibdata-qa1.princeton.edu]: FAILED! => {"changed": false, "msg": "unable to connect to database: connection to server at \"lib-postgres-staging3.princeton.edu\" (128.112.204.72), port 5432 failed: FATAL:  password authentication failed for user \"postgresadmin\"\nconnection to server at \"lib-postgres-staging3.princeton.edu\" (128.112.204.72), port 5432 failed: FATAL:  password authentication failed for user \"postgresadmin\"\n"}
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: connection to server at "lib-postgres-staging3.princeton.edu" (128.112.204.72), port 5432 failed: FATAL:  password authentication failed for user "postgresadmin"
fatal: [bibdata-worker-qa1.princeton.edu]: FAILED! => {"changed": false, "msg": "unable to connect to database: connection to server at \"lib-postgres-staging3.princeton.edu\" (128.112.204.72), port 5432 failed: FATAL:  password authentication failed for user \"postgresadmin\"\nconnection to server at \"lib-postgres-staging3.princeton.edu\" (128.112.204.72), port 5432 failed: FATAL:  password authentication failed for user \"postgresadmin\"\n"}
```

I wonder if it is related to the postgres migration?  @kayiwa?

carolyncole commented:
This PR is actually better than running the playbook in main.  It fails at pg_hba in main.  
```
TASK [postgresql : PostgreSQL | ensure access to postgres server for client connections] **************************************
fatal: [bibdata-worker-qa1.princeton.edu -> lib-postgres3.princeton.edu]: FAILED! => {"changed": false, "msg": "Destination /etc/postgresql/13/main/pg_hba.conf does not exist !", "rc": 257}
fatal: [bibdata-qa1.princeton.edu -> lib-postgres3.princeton.edu]: FAILED! => {"changed": false, "msg": "Destination /etc/postgresql/13/main/pg_hba.conf does not exist !", "rc": 257}
```
